ctapipe/io/tableloader.py

```diff
@@ -0,0 +1,330 @@
+
+        self.instrument_table = None
+        if self.load_instrument:
+            table = self.subarray.to_table()
+            optics = self.subarray.to_table(kind="optics")
+            optics["optics_index"] = np.arange(len(optics))
+            optics.remove_columns(["name", "description", "type"])
+            table = join(
+                table,
+                optics,
+                keys="optics_index",
+                # conflicts for TAB_VER, TAB_TYPE, not needed here, ignore
+                metadata_conflicts="silent",
+            )
+
+            table.remove_columns(["optics_index", "camera_index"])
+            self.instrument_table = table
+
+        self.shower_table = None
+        if self.load_simulated and SHOWER_TABLE in self.h5file:
+            self.shower_table = read_table(self.h5file, SHOWER_TABLE)
+
+    def close(self):
+        self.h5file.close()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def _get_reconstructors(self):
+
+        if GEOMETRY_GROUP in self.h5file:
+            g = self.h5file.root[GEOMETRY_GROUP]
+        else:
+            raise ValueError(f"No DL2 geometry data in h5file: {self.h5file}")
+        
+        reconstructors = g._v_children.keys()
+        
+        return reconstructors
+
+    def _read_telescope_table(self, group, tel_id):
+
+        if self.structure == "by_id":
+            key = f"{group}/tel_{tel_id:03d}"
+            condition = None
+        else:
+            key = f"{group}/{self.subarray.tel[tel_id]!s}"
+            condition = f"tel_id == {tel_id}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=condition)
+        else:
+            table = Table()
+
+        return table
+    
+    def _read_telescope_DL2_geometry(self, group, reconstructor):
+        
+        key = f"{group}/{reconstructor}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=None)
+        else:
+            table = Table()
+
+        return table
+
+    def read_subarray_events(self, labels=None, sort_by=["obs_id", "event_id"]):
+        """Read data from file.
+        
+        Parameters
+        ----------
+        labels: list
+            Either tel_ids or tel_types depending on dataset split.
+        sort_by: list
+            Columns to sort the final table.
+            By default it is sorted by obs_id and event_id.
+        
+        Returns
+        -------
+        output: astropy.Table
+            Single table containing the enabled data information."""
+        
+        if not any(self.all_loads):
+            self.close()
+            raise ValueError("No data enabled for loading.")
+
+        if self.load_instrument and (not any([self.load_dl1_images, self.load_dl1_parameters, self.load_true_images])):
```

Comment:
In my implementation, `load_instrument` was an option to enable joining instrument info onto the other information. I think this makes sense, since if someone is only interested in instrument information, we have `SubarrayDescription` also accessible from this loader.

Reply:
sure this is still possible, it's just something more I felt we should add if you allow to read any single table

Reply:
We can probably drop the load_instrument option: it's only for a very specific use case when doing some debugging plots, and the info is already in the subarray description.  The user can always join it manually if needed - it adds a lot of redundant information (e.g. the telescope position for each event), which normally you don't need.

Reply:
that I would be in favor of (since we have `SubarrayDescription` already) but maybe it's Max should say if he wants to keep it since he added it in the first place, maybe he had some reason add this support here

Reply:
of course you want it together with some other info like he said above, but if you put it as its separate traitlet I (as a user) would expect to be able to export it alone if I wish

another option is to always attach it (are there cases in which we do NOT expect this info to be in a file?)

Reply:
> We can probably drop the load_instrument option: it's only for a very specific use case when doing some debugging plots, and the info is already in the subarray description. 

No, this is also useful for e.g. building a single machine learning model for multiple telescope types (using focal length / mirror area as features)

Reply:
as it is now one can load either only instrument data or attach it to a table containing `tel_id` as a column name (so if at least one between `load_dl1_images`, `load_dl1_parameters` and `load_true_images` has been set to True) resulting in a raised `ValueError` if only `event_id`-based data is loaded with instrument data

Reply:
this disussion is still relevant I guess